WordPress/openverse
# Deduce Python/Node.js/pnpm version in Docker images

## Fixes

Fixes #409

## Description




This PR takes inspiration from the catalog to automatically use the right Python version in the Docker image. To do this, take the example of the API.

- `api/justfile` exports `$API_PY_VERSION` which is derived by transforming the `python_version` field in the `Pipfile`.
- `api/justfile` provides a recipe `py-version` that echoes `$API_PY_VERSION`.
- `justfile` exports `$API_PY_VERSION` by invoking `api/py-version`.
- `docker-compose.yml` maps build arg API_PY_VERSION to `$API_PY_VERSION` env var.
- `api/Dockerfile` uses directive `FROM python:${API_PY_VERSION}` to automatically derive from the right Python image.

The same goes for the ingestion server. This extends to the frontend (w.r.t Node and pnpm versions) to implement the cleaner solution suggested in PR #1051.

## Testing Instructions

Please run the following just recipes and check their output.

- just versions (covers all the ones below)
- just catalog/py-version
- just catalog/airflow-version
- just api/py-version
- just ingestion_server/py-version
- just frontend/node-version
- just frontend/pnpm-version

## Checklist



      `Update index.md`).
      a parent feature branch.
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin



<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
